### PR TITLE
byte_array_impl is io_blob_safe with RAII and movable

### PR DESCRIPTION
**Pull request by hkadayam**

### CODE REVIEW

**Review by szmyd — CHANGES_REQUESTED** [review]
Source branch is based on 9.x; either rebase or we can merge 9.x into master after this gets re-targeted.

File: include/sisl/fds/buffer.hpp [hunk 1]
```diff
@@ -300,24 +301,48 @@ struct io_blob : public blob {
...
+
+    io_blob_safe(io_blob_safe const& other) = delete;
+    io_blob_safe(io_blob_safe&& other) : io_blob{other.bytes, other.size, other.is_aligned} {
+        other.bytes = nullptr;
+        other.size = 0;
+    }
+
+    io_blob_safe& operator=(io_blob_safe const& other) = delete; // Delete copy constructor
+    io_blob_safe& operator=(io_blob_safe&& other) {
+        if (bytes != nullptr) { this->buf_free(m_tag); }
```
Comment (szmyd) [inline include/sisl/fds/buffer.hpp:324]: Why not swap and let the caller's scope take care of it?

**Review by szmyd — CHANGES_REQUESTED** [review]

File: include/sisl/fds/buffer.hpp (hunk 1 above)
Comment (hkadayam) [reply include/sisl/fds/buffer.hpp:324]: Can we technically swap, given the semantics of std::move() is that post move, the moved entity should be reset? I am worried, if someone uses the value after move, would give indifferent result

```
io_blob_safe b1{ptr1, 1024};
io_blob_safe b2{ptr2, 10};
auto b2 = std::move(b);
...
...
std::cout << b.size << "\n" // Shouldn't this required to be report 0 and b.bytes have nullptr?
```

File: include/sisl/fds/buffer.hpp [hunk 2]
```diff
@@ -300,24 +301,48 @@ struct io_blob : public blob {
...
+public:
+    io_blob_safe(uint32_t sz, uint32_t alignment = 0, buftag tag = buftag::common) :
+            io_blob(sz, alignment, tag), m_tag{tag} {}
+    io_blob_safe(uint8_t* bytes, uint32_t size, bool is_aligned) : io_blob(bytes, size, is_aligned) {}
+    virtual ~io_blob_safe() {
+        if (bytes != nullptr) { io_blob::buf_free(m_tag); }
+    }
+
+    io_blob_safe(io_blob_safe const& other) = delete;
+    io_blob_safe(io_blob_safe&& other) : io_blob{other.bytes, other.size, other.is_aligned} {
```
Comment (szmyd) [inline include/sisl/fds/buffer.hpp:317]: `other.aligned`

File: include/sisl/fds/buffer.hpp (hunk 1 above)
Comment (szmyd) [reply include/sisl/fds/buffer.hpp:324]: No, i think it's meant to be indeterministic though.

**Pull request closed without merging**